### PR TITLE
Python: Implement select

### PR DESCRIPTION
**Pull request by Fokko**
So we can select a subset of the fields in a schema

### CODE REVIEW

File: python/pyiceberg/exceptions.py [hunk 1]
```diff
@@ -82,3 +82,7 @@ class NoSuchPropertyException(Exception):
 
 class NotInstalledError(Exception):
     """When an optional dependency is not installed"""
+
+
+class ColumnNotFoundError(Exception):
+    """When selecting a column that doesn't exist"""
```
Comment (rdblue) [inline python/pyiceberg/exceptions.py:88]: In Java, we just throw IllegalArgumentException (equivalent to ValueError, I think) when something can't be found.

File: python/tests/test_schema.py [hunk 2]
```diff
@@ -671,3 +672,24 @@ def test_prune_columns_struct_in_map_full():
...
     ids = set(range(table_schema_nested.highest_field_id))
     assert prune_columns(table_schema_nested, ids, True) == table_schema_nested
+
+
+def test_schema_select(table_schema_nested: Schema):
+    assert table_schema_nested.select("bar", "baz") == Schema(
+        NestedField(field_id=2, name="bar", field_type=IntegerType(), required=True),
+        NestedField(field_id=3, name="baz", field_type=BooleanType(), required=False),
+        schema_id=1,
+        identifier_field_ids=[1],
```
Comment (rdblue) [inline python/tests/test_schema.py:682]: Minor: Should this be updated? This schema doesn't have all of the identifier field IDs, so maybe we should clear the identifier field ID list?

File: python/pyiceberg/scan.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+# specific language governing permissions and limitations
+# under the License.
+from typing import Optional, Union
+
+from pyiceberg.expressions.base import BooleanExpression
+from pyiceberg.schema import Schema
+from pyiceberg.table import Snapshot, Table
+
+
+class TableScan:
```
Comment (rdblue) [inline python/pyiceberg/scan.py:24]: Can we move this to a separate PR? I'm not sure why it is in a PR that is adding `select` support.

File: python/pyiceberg/exceptions.py (hunk 1 above)
Comment (Fokko) [reply python/pyiceberg/exceptions.py:88]: I figured this would be a nice one to add since it is very user facing. We're throwing a lot of `ValueError`'s already 😱

File: python/tests/test_schema.py (hunk 2 above)
Comment (Fokko) [reply python/tests/test_schema.py]: Great catch, I'll now take the intersect of the `identifier_field_ids` and the IDs of the fields

File: python/pyiceberg/schema.py [hunk 4]
```diff
@@ -800,7 +800,11 @@ def primitive(self, primitive: PrimitiveType) -> PrimitiveType:
 
 def prune_columns(schema: Schema, selected: Set[int], select_full_types: bool = True) -> Schema:
     result = visit(schema.as_struct(), _PruneColumnsVisitor(selected, select_full_types))
-    return Schema(*(result or StructType()).fields, schema_id=schema.schema_id, identifier_field_ids=schema.identifier_field_ids)
+    return Schema(
+        *(result or StructType()).fields,
+        schema_id=schema.schema_id,
+        identifier_field_ids=list(selected.intersection(schema.identifier_field_ids)),
```
Comment (rdblue) [inline python/pyiceberg/schema.py:806]: I don't think `intersection` is correct. The identifier field IDs are the set of fields that make a unique key. The resulting schema either has the same identifier fields (i.e. the same key) or doesn't have a key at all. I don't think it should ever have a different identifier/key.

So I think this should be `schema.identifier_field_ids if all(id in selected for id in schema.identifier_field_ids) else []`.

**Review by rdblue — APPROVED** [review]
One minor issue with the identifier ID handling, but overall this is great. Thanks, @Fokko!
